mountpoint-s3/src/data_cache/express_data_cache.rs

```diff
@@ -148,16 +167,35 @@ where
 
         let object_key = block_key(&self.prefix, &cache_key, block_idx);
 
-        // TODO: ideally we should use a simple Put rather than MPU.
-        let params = PutObjectParams::new();
-        let mut req = self
+        let mut params = PutObjectSingleParams::new();
+        let (sse_type, key_id) = self
+            .sse
+            .clone()
+            .into_inner()
+            .map_err(|err| DataCacheError::IoFailure(err.into()))?;
+        params = params.server_side_encryption(sse_type);
+        params = params.ssekms_key_id(key_id);
+
+        let (data, _crc) = bytes.into_inner().map_err(|_| DataCacheError::InvalidBlockContent)?;
+        let req = self
             .client
-            .put_object(&self.bucket_name, &object_key, &params)
+            .put_object_single(&self.bucket_name, &object_key, &params, data)
             .in_current_span()
             .await?;
-        let (data, _crc) = bytes.into_inner().map_err(|_| DataCacheError::InvalidBlockContent)?;
-        req.write(&data).await?;
-        req.complete().await?;
+
+        // Verify that headers of the PUT response match the expected SSE
+        if let Err(err) = self
+            .sse
+            .verify_response(req.sse_type.as_deref(), req.sse_kms_key_id.as_deref())
+        {
+            error!(key=?cache_key, error=?err, "A cache block was stored with wrong encryption settings");
+            // Reaching this point is very unlikely and means that SSE settings were corrupted in transit or on S3 side, this may be a sign of a bug
+            // in CRT code or S3. Thus, we terminate Mountpoint to send the most noticeable signal to customer about the issue. We prefer exiting
+            // instead of returning an error because:
+            // 1. this error would only be reported to logs because the cache population is an async process
+            // 2. the reported error is severe as the object was already uploaded to S3.
+            std::process::exit(1);
```

Comment:
> S3 told us it ignored us
Yes, only in that case, which may also happen if CRT stops adding those headers to PutObject requests to S3 Express.

> Can we hit this path if the bucket's encryption settings change after startup?
No, in this case we expect to bail out with `PutObjectSingle` error at line [157](https://github.com/awslabs/mountpoint-s3/pull/1131/files#diff-9dcd97550fdeced58b279050424f1389af943edc9f652a4514393f7804be9420L157). We provide the settings that S3 is expected to enforce. If it is not able to do that (because the default for the bucket is now different), the PutObject will fail.